Add NOLO controller model to SteamVR

johnlajoie opened the pull request:
This pull request adds a SteamVR rendermodel for the NOLO controllers and sets OSVRTrackedController to use it by default. The model is based on the model released by LYRobitix: 

https://github.com/NOLOVR/NOLO-Others/tree/master/NOLO_Controller_Model

I had to use Blender to modify the structure of the model and the texturing so SteamVR would accept it. The Blender project file is also included in case anyone wants to improve from here. 

The resources/rendermodels/nolo_controller directory should be copied to the SteamVR installation directory under

Steam/steamapps/common/SteamVR/resources/rendermodels

Commit:
add NOLO controller model for NOFC

Commit:
add OBJ file as binary

Commit:
added controllerRenderModel to default.vrsettings

nanospork commented:
I don’t think that SteamVR directory should be necessary. I think resources can be provided alongside the driver, but I’m not sure how.

johnlajoie commented:
This is how the OSVR camera model is provided, for example.  The latest commits added the configuration property "controllerRenderModel" so that the specific model is not hard-coded in the driver, but can be changed by editing settings/default.vrsettings.  Of course, the default in default.vrsettings is the nolo controller model. 

Are you suggesting the model be included with the driver binary? That seems to go counter to the OSVR and OpenVR design philosophy, and it isn't very flexible.  In the future, if there is some new controller hardware that someone wants to include they won't need to recompile SteamVR, just add the new model and change the configuration file. 

So... I disagree. I think this is the most flexible and general way to include controller models in SteamVR. 

(BTW - the new changes make obsolete the previous comments about copying the model to the Steam directory. It now resides as part of the SteamVR directory, just like the camera model.)

My latest commits and your comments may have crossed? As I read your comments again maybe this is exactly what you meant by "alongside the driver"?

Commit:
new model with functional trackpad, trigger, buttons

johnlajoie commented:
I have updated the NOLO model to include functional components, similar to the Vive controller model.  The trigger moves, the buttons (system, menu, trackpad, grip) move when pressed,  The trackpad also shows the little touch marker when touched, 

There are some differences.  I really don't like the menu scroll wheel that shows up superimposed on the Vive controller model trackpad when you are on an active menu, so I left the trackpad touch marker active instead.  I also haven't figured out the status panel yet so I don't have the lights indicating battery charge either.

To clarify the previous discussion - there is no longer any need to copy this to the SteamVR directory. Just include it in the rendermodels directory for the SteamVR-OSVR driver and default.vrsettings will have the driver pick up the correct model. 

I think it would be very nice to include this in NOFC.

Commit:
couldn't help myself, had to make the LED green

Commit:
fix velocity and angular velocity not getting through to SteamVR

Commit:
add full range of configuration options

Commit:
explicit conversion from quaternion to axis angle

johnlajoie commented:
I can see this pull request is getting contaminated with my latest work trying to get the velocity and acceleration data recognized in SteamVR - I should have created a new branch.  

Here's an FYI one what I've been doing: 

I found multiple issues with velocity information being passed to SteamVR - the most ugly was that SteamVR wasn't actually getting the linear velocity information at all! osvrGetVelocityState was giving back a return value that looked like an uninitialized value, so the check was failing and no velocity information was actually being passed to SteamVR.  

In addition, I found that the axis-angle conversion to quaternion in nolo-osvr was wrong, so when it was converted back in OSVR-SteamVR it was a mess.  I fixed this by being very explicit about the conversions, even if it isn't super-pretty code.  Along the way I added additional flags to SteamVR so the linear and angular components of the acceleration and velocity can be controlled by the settings file. 

Along the way there were also changes to OSVR-fusion to pass along the angular data. 

In addition, I've been following along with the source code nolo posted for their OpenVR driver to make certain the end result is that we are passing in the information through our chain in a way that it ends up in the same state. 

As of this morning I have tested the linear and angular velocity information and it's no longer giving me crazy values. If I only turn on the linear velocity the controllers jitter a lot, but when I turn on the angular velocity as well this jitter is greatly reduced.  I haven't tested the acceleration data yet, but I note that the nolo driver doesn't pass it along.

I wish I could say this was all a great success, but I still can't throw a stick to the mechanical dog in "The Lab". I have the feeling this is related to my testing being in ceiling mode - we transform the position vector using the server config file, but not the velocity or acceleration data.  I will test in "normal" mode later today and dig into this further.  It may be that to properly support ceiling mode we may have to go back to using the nolo API for this and having some configuration flag for nolo-osvr.

Commit:
add linear velocity scaling to allow testing and tuning

Commit:
latest updates and debugging

Commit:
separate out flags for HMD and controllers

johnlajoie commented:
IT FINALLY WORKS!!

There were *many* reasons that the velocity data was not getting to SteamVR - the last one that took me forever to figure out was that the config files didn't have the fusion controller timestamped, so it was only sending up position information. 

With the latest in my NOFC, nolo_osvr, OSVR-fusion and SteamVR-OSVR repositories I can finally throw a stick to a mechanical dog in "The Lab".  That was one hell of a lot of work to just throw a stick. Controller motion is nice and smooth with no jitters. (This was all in "horizontal" mode.)

So far I have only tested the linear velocity data. I would like to at least test the angular velocity data as well as that can be important for throwing too. The acceleration data is all there up to the SteamVR driver, the default configuration just ignores it. 

TO DO: 
-> Test at least angular velocity in horizontal mode.  If time permits, test the acceleration data as well, although I am not sure SteamVR uses it? 
-> Go back to work on ceiling mode.  I don't think that using the server configuration file to transform the data will work because I will also need to transform the linear velocity data.  I suspect that I will either need to add a configuration parameter to the windows version of nolo_osvr to enable ceiling mode in the LyRobotix API, or we will need to distribute to different version of the nolo-osvr driver. 

I'm hopeful I can get this done this weekend.  If you're up for it, at that point I think a new release would be in order.

nanospork commented:
I had velocity data working as of my last commit, including the time stamp fix... did something change? Did I miss something?

Sent from my iPhone

> On Feb 14, 2018, at 6:51 PM, John Lajoie <notifications@github.com> wrote:
> 
> IT FINALLY WORKS!!
> 
> There were many reasons that the velocity data was not getting to SteamVR - the last one that took me forever to figure out was that the config files didn't have the fusion controller timestamped, so it was only sending up position information.
> 
> With the latest in my NOFC, nolo_osvr, OSVR-fusion and SteamVR-OSVR repositories I can finally throw a stick to a mechanical dog in "The Lab". That was one hell of a lot of work to just throw a stick. Controller motion is nice and smooth with no jitters. (This was all in "horizontal" mode.)
> 
> So far I have only tested the linear velocity data. I would like to at least test the angular velocity data as well as that can be important for throwing too. The acceleration data is all there up to the SteamVR driver, the default configuration just ignores it.
> 
> TO DO:
> -> Test at least angular velocity in horizontal mode. If time permits, test the acceleration data as well, although I am not sure SteamVR uses it?
> -> Go back to work on ceiling mode. I don't think that using the server configuration file to transform the data will work because I will also need to transform the linear velocity data. I suspect that I will either need to add a configuration parameter to the windows version of nolo_osvr to enable ceiling mode in the LyRobotix API, or we will need to distribute to different version of the nolo-osvr driver.
> 
> I'm hopeful I can get this done this weekend. If you're up for it, at that point I think a new release would be in order.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

johnlajoie commented:
Nice to hear from you!  As for things working with the 0.3rc1 release, I don't believe it.   I see that in the server config files that *one* of the four config files tagged for the release has the controller data timestamped. But there were other issues as well that were preventing the velocity data from being used in SteamVR - like the linearVelocityValid flags not being set in the fusion driver, so SteamVR never used the data.

johnlajoie commented:
BTW - I tested using the Lab because most applications will try to guess velocity from position information if velocity is not provided.  You can still throw, but the performance is lackluster at best.  DoomVFR does this, for example.  However, the Lab is intended as a developers tool so it explicitly does *not* do this - no velocity, no throw.  I started down this road because I didn't find throwing in general any better in v0.3RC1 and when I tested in the Lab that's what gave me the clue that velocity information wasn't getting through the driver chain. I think this might be why we see comments that throwing still feels weak with v0.3RC1.

As an aside - so people can tune this to their preference I added a setting in the config file that will scale the linear velocity.  It's set at 1.0 by default and IMHO doesn't seem to be needed, but I thought it was a nice configuration option. I saw that the psmove folks had something similar.

nanospork commented:
I only ever actually used the one config, so that might explain it. But I definitely had velocity working in RecRoom, which does not guess at velocity if it’s not provided (when you tried to throw before RC1, whatever you let go of would drop straight down. With RC1, it was “weak” but it worked.)

I’ll look over your code, it may be cleaner than my implementation anyway and it looks like you’ve added a lot of great options!

> On Feb 15, 2018, at 7:41 AM, John Lajoie <notifications@github.com> wrote:
> 
> BTW - I tested using the Lab because most applications will try to guess velocity from position information if velocity is not provided. You can still throw, but the performance is lackluster at best. DoomVFR does this, for example. However, the Lab is intended as a developers tool so it explicitly does not do this - no velocity, no throw. I started down this road because I didn't find throwing in general any better in v0.3RC1 and when I tested in the Lab that's what gave me the clue that velocity information wasn't getting through the driver chain. I think this might be why we see comments that throwing still feels weak with v0.3RC1.
> 
> As an aside - so people can tune this to their preference I added a setting in the config file that will scale the linear velocity. It's set at 1.0 by default and IMHO doesn't seem to be needed, but I thought it was a nice configuration option. I saw that the psmove folks had something similar.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

johnlajoie commented:
Hmmm.  Well, I definitely could have had something screwed up when I started, or just confused myself.  Wouldn't be the first time.  I'd be happy to hear what you think, appreciate your time.

nanospork commented:
It was probably the time stamp thing mainly. It took me forever to find that. I thought I made a note of it somewhere, but I must have forgotten - sorry you had to go through that same bug hunt!

> On Feb 15, 2018, at 7:49 AM, John Lajoie <notifications@github.com> wrote:
> 
> Hmmm. Well, I definitely could have had something screwed up when I started, or just confused myself. Wouldn't be the first time. I'd be happy to hear what you think, appreciate your time.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

johnlajoie commented:
When you have a moment I could use your advice....

The NOLO CeilingMode API doesn't seem to work - none of the coordinates seem to be transformed.  If I just avoid this and set a flag in the driver to do the transformation, ceiling mode works great (including throwing).  I think we need to do this transformation in the driver now since we can't transform the velocity and acceleration from the server config file. 

So that leaves me a with a question - is there a way to get a parameter to an OSVR hardware plugin? From looking though the OSVR API there doesn't seem to be - unlike a driver plugin it doesn't get called with a parameter that gives it access to the Json data in the server config file.  Without this we'll just need to deliver two compiled versions of the the nolo-osvr plugin? 

Here's a little update - ceiling mode works great (with the changes above), and the angular velocity works great too after some fixes. You can now toss something by flicking your wrist in place.  I need to do a bit more testing with the acceleration values, but if the data is good they should work properly.

johnlajoie commented:
... and while you're at it, here's another question.  Why do I have to re-run room setup every time I start SteamVR? The old Nolo software used to overwrite SteamVR configurations so I though maybe the SDK is still doing this, but I couldn't find it in a quick look through the DLL.  If it's there I'd love to patch the DLL and get rid of this behavior. DO you have this issue as well? 

I have updated by NOFC directory with the latest build and I'm keeping it current in case you're interested. I've edited the README's a bit so that someone won't confuse it with your (official) repo.

johnlajoie commented:
In  my NOFC repository I just checked in a version of NoLo_USBHID.dll that has been modified so that it no longer edits the SteamVR chaperone_info.vrchap file. That was a really nasty piece of work by the Nolo folks.

Commit:
cleanups

johnlajoie commented:
After testing I'm satisfied that the linear velocity and acceleration are working just fine so I have enabled them by default, and I've set the default linear velocity multiplier to 1.5.  To meet that makes throwing feel much more natural. My NOFC repo is up to date with the latest documentation and builds. 

I'm not happy with the angular quantities.  If I enable the angular velocity, the controllers flash and flicker, indicating that SteamVR is unhappy with something. The "flick of the wrist" still seems to work like I would expect, but a normal throw is messed up.  I'm going to keep digging on this. 

UPDATE: I take back what I said about acceleration - the linear acceleration wasn't getting through due to a bad check in the SteamVR driver; that's fixed.  However, I find that when anything but the linear velocity is allowed through (and actually gets through), SteamVR Home won't start.  My NOFC repository is back to velocity only for the time being and the documentation is updated.

Commit:
configure flags before setting callbacks

Commit:
cleanup log message

Commit:
fix return code check for acceleration reports

Commit:
fixed angular quantities

johnlajoie commented:
I think it's done. SteamVR really did want the angular quantities in the global frame, and it seems to get pissed (SteamVR Home crashes) if you have linear acceleration but no angular velocity.  I can now enable all four quantities at once and everything works quite nicely, however, with the accelerations on the controllers are slightly jittery. 

My NOFC repo is updated along with the accompanying documentation. Give it a shot, I'd like to know how it works for you.

johnlajoie committed:
Create README.md
